tokenize_rt.py

```diff
@@ -6,35 +6,65 @@
 import io
 import re
 import tokenize
+from typing import Iterable
+from typing import Iterator
+from typing import List
+from typing import Optional
+from typing import Pattern
+from typing import Sequence
+from typing import Tuple
+
+if False:
+    from typing import Text
 
 
 ESCAPED_NL = 'ESCAPED_NL'
 UNIMPORTANT_WS = 'UNIMPORTANT_WS'
 NON_CODING_TOKENS = frozenset(('COMMENT', ESCAPED_NL, 'NL', UNIMPORTANT_WS))
-Offset = collections.namedtuple('Offset', ('line', 'utf8_byte_offset'))
-Offset.__new__.__defaults__ = (None, None)
-Token = collections.namedtuple(
+
+
+class Offset(collections.namedtuple('Offset', ('line', 'utf8_byte_offset'))):
+    __slots__ = ()
+
+    def __new__(cls, line=None, utf8_byte_offset=None):
+        # type: (Optional[int], Optional[int]) -> None
+        return super(Offset, cls).__new__(cls, line, utf8_byte_offset)
+
+
+class Token(collections.namedtuple(
     'Token', ('name', 'src', 'line', 'utf8_byte_offset'),
-)
-Token.__new__.__defaults__ = (None, None)
-Token.offset = property(lambda self: Offset(self.line, self.utf8_byte_offset))
+)):
+    __slots__ = ()
+
+    def __new__(cls, name, src, line=None, utf8_byte_offset=None):
+        # type: (Text, Text, Optional[int], Optional[int]) -> None
+        return super(Token, cls).__new__(
+            cls, name, src, line, utf8_byte_offset
+        )
+
+    @property
+    def offset(self):  # type: () -> Offset
+        return Offset(self.line, self.utf8_byte_offset)
+
 
 _string_re = re.compile('^([^\'"]*)(.*)$', re.DOTALL)
 _string_prefixes = frozenset('bfru')
 _escaped_nl_re = re.compile(r'\\(\n|\r\n|\r)')
 
 
 def _re_partition(regex, s):
+    # type: (Pattern[Text], Text) -> Tuple[Text, Text, Text]
     match = regex.search(s)
     if match:
         return s[:match.start()], s[slice(*match.span())], s[match.end():]
     else:
         return (s, '', '')
 
 
-def src_to_tokens(src):
+def src_to_tokens(src):  # type: (Text) -> List[Token]
     tokenize_target = io.StringIO(src)
-    lines = (None,) + tuple(tokenize_target)
+    lines = ('',) + tuple(tokenize_target)
```

Comment:
hmmmm so this was actually intentional so a programming error would trigger a `TypeError` -- but I guess this is fine too 🤷‍♂

Reply:
I see. I thought it was so you didn't have to subtract 1 in the indexing strategy below. Typing becomes annoying without this change 🤷‍♂

Reply:
it's also for that, but if I made a mistake I'd get a good error

if I wanted to keep that I could

```python
class NoZeroOnlyPositive:
    def __init__(self, seq):
        self.seq = seq
    def __getitem__(self, ind):
        if ind == 0:
            raise TypeError('no!')
       else:
           return self.seq[ind]
```